File: Code Examples/Security_OTP_Example/src/main/java/io/hashimati/security/services/CustomRefreshTokenPersistence.java
package io.hashimati.security.services;


import io.hashimati.security.domains.RefreshToken;
import io.hashimati.security.repository.RefreshTokenRepository;
import io.micronaut.security.authentication.Authentication;
import io.micronaut.security.errors.OauthErrorResponseException;
import io.micronaut.security.token.event.RefreshTokenGeneratedEvent;
import io.micronaut.security.token.refresh.RefreshTokenPersistence;
import jakarta.inject.Singleton;
import org.reactivestreams.Publisher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Flux;
import reactor.core.publisher.FluxSink;

import java.util.Optional;

import static io.micronaut.security.errors.IssuingAnAccessTokenErrorCode.INVALID_GRANT;

@Singleton
public class CustomRefreshTokenPersistence implements RefreshTokenPersistence {
    private final RefreshTokenRepository refreshTokenRepository;
    private static final Logger log = LoggerFactory.getLogger(CustomRefreshTokenPersistence.class);

    public CustomRefreshTokenPersistence(RefreshTokenRepository refreshTokenRepository) {
        this.refreshTokenRepository = refreshTokenRepository;
    }

    @Override
    public void persistToken(RefreshTokenGeneratedEvent event) {
        log.info("Persist Token by {}", event.getAuthentication().getName());
        if (event != null &&
                event.getRefreshToken() != null &&
                event.getAuthentication() != null &&
                event.getAuthentication().getName() != null) {
            String payload = event.getRefreshToken();
            RefreshToken refreshToken = new RefreshToken();
            refreshToken.setRefreshToken(payload);
            refreshToken.setUsername(event.getAuthentication().getName());
            refreshToken.setRevoked(false);
            refreshTokenRepository.save(refreshToken);
        }
    }

    @Override
    public Publisher<Authentication> getAuthentication(String refreshToken) {
        log.info("Get Authentication by refreshToken:{}", refreshToken);
        return Flux.create(emitter -> {
            Optional<RefreshToken> tokenOpt = refreshTokenRepository.findByRefreshToken(refreshToken);
            if (tokenOpt.isPresent()) {
                RefreshToken token = tokenOpt.get();
                if (token.getRevoked()) {
                    emitter.error(new OauthErrorResponseException(INVALID_GRANT, "refresh token revoked", null));
                } else {
                    emitter.next(Authentication.build(token.getUsername()));
                    emitter.complete();
                }
            } else {
                emitter.error(new OauthErrorResponseException(INVALID_GRANT, "refresh token not found", null));
            }
        }, FluxSink.OverflowStrategy.ERROR);
    }
}

File: MicroCli/src/main/resources/micronaut/security/jwt/kotlin/JDBC/services/CustomRefreshTokenPersistence.kt
package ${securityPackage}.services

import ${securityPackage}.domains.RefreshToken
import ${securityPackage}.repository.RefreshTokenRepository
import ${securityPackage}.authentication.Authentication
import io.micronaut.security.errors.IssuingAnAccessTokenErrorCode
import io.micronaut.security.errors.OauthErrorResponseException
import io.micronaut.security.token.event.RefreshTokenGeneratedEvent
import io.micronaut.security.token.refresh.RefreshTokenPersistence
import jakarta.inject.Singleton
import org.reactivestreams.Publisher
import org.slf4j.LoggerFactory
import reactor.core.publisher.Flux
import reactor.core.publisher.FluxSink
import java.time.Instant


@Singleton
class CustomRefreshTokenPersistence(private val refreshTokenRepository: RefreshTokenRepository) :
    RefreshTokenPersistence {
    override fun persistToken(event: RefreshTokenGeneratedEvent) {
        log.info("Persist Token by {}", event.authentication.name)
        if ( event.refreshToken != null && event.authentication != null && event.authentication.name != null) {
            val payload = event.refreshToken
            val refreshToken = RefreshToken(1, "", "", false, Instant.now());
            refreshToken.refreshToken = payload
            refreshToken.username = event.authentication.name
            refreshToken.revoked = false
            refreshTokenRepository.save(refreshToken)
        }
    }

    override fun getAuthentication(refreshToken: String): Publisher<Authentication> {
        log.info("Get Authentication by refreshToken:{}", refreshToken)
        return Flux.create({ emitter: FluxSink<Authentication> ->
            val tokenOpt = refreshTokenRepository.findByRefreshToken(refreshToken)
            if (tokenOpt!!.isPresent) {
                val (_, username, _, revoked) = tokenOpt.get()
                if (revoked) {
                    emitter.error(
                        OauthErrorResponseException(
                            IssuingAnAccessTokenErrorCode.INVALID_GRANT,
                            "refresh token revoked",
                            null
                        )
                    )
                } else {
                    emitter.next(Authentication.build(username))
                    emitter.complete()
                }
            } else {
                emitter.error(
                    OauthErrorResponseException(
                        IssuingAnAccessTokenErrorCode.INVALID_GRANT,
                        "refresh token not found",
                        null
                    )
                )
            }
        }, FluxSink.OverflowStrategy.ERROR)
    }

    companion object {
        private val log = LoggerFactory.getLogger(CustomRefreshTokenPersistence::class.java)
    }
}
